### PR TITLE
Make Store.Flush() guarantee immediate flush on sinks with that support

### PR DESCRIPTION
**Pull request by jsedgwick**
Attempt at fixing bug demonstrated in #39 
@mjchoi can you patch this and see what it does for you?

### CODE REVIEW

**Comment by jsedgwick** [comment]
PTAL @btc @rodaine @mjchoi

**Comment by mjchoi** [comment]
just tested this patch.  still seeing early termination.

```
root@obstools-legacy:/code/obstools# ./devbox-helper-gcpscraper.sh
{"level":"debug","msg":"sinkWriter.Write() wrote to outc chan","time":"2018-07-30T23:36:26Z"}
{"level":"debug","msg":"tcpStatsdSink.Flush() returning","time":"2018-07-30T23:36:26Z"}
{"level":"debug","msg":"statStore.Flush() returning","time":"2018-07-30T23:36:26Z"}
```

with sleep:
```
root@obstools-legacy:/code/obstools# ./devbox-helper-gcpscraper.sh
{"level":"debug","msg":"sinkWriter.Write() wrote to outc chan","time":"2018-07-30T23:37:00Z"}
{"level":"debug","msg":"tcpStatsdSink.Flush() returning","time":"2018-07-30T23:37:00Z"}
{"level":"debug","msg":"tcpStatsdSink.run() received from outc chan","time":"2018-07-30T23:37:00Z"}
{"level":"debug","msg":"statStore.Flush() returning","time":"2018-07-30T23:37:00Z"}
{"level":"debug","msg":"tcpStatsdSink.run() wrote to conn","time":"2018-07-30T23:37:00Z"}
{"level":"debug","msg":"tcpStatsdSink.Flush() returning","time":"2018-07-30T23:37:01Z"}
{"level":"debug","msg":"tcpStatsdSink.Flush() returning","time":"2018-07-30T23:37:01Z"}
```

**Comment by mjchoi** [comment]
thanks for taking a look! 😄

**Comment by jsedgwick** [comment]
ah sorry i did not see the full extent of the issue, also need to block on the actual flush happening if near termination of a program like this, willfix tomorrow. thx for checking @mjchoi

File: logging_sink.go [hunk 1]
```diff
@@ -20,3 +20,7 @@ func (s *loggingSink) FlushGauge(name string, value uint64) {
 func (s *loggingSink) FlushTimer(name string, value float64) {
 	logger.Debugf("[gostats] flushing time %s: %f", name, value)
 }
+
+func (s *loggingSink) Flush() {
+	logger.Debugf("[gostats] Flush() called, all stats would be flushed")
```
Comment (btc) [inline logging_sink.go]: i see how this verb tense is accurate, but it reads a little strangely. 🤷‍♂️

**Comment by btc** [comment]
PTAL @jsedgwick

**Comment by jsedgwick** [comment]
PTAL @mjchoi can you try this?

**Comment by mjchoi** [comment]
sure.  will report back

**Comment by mjchoi** [comment]
same as before:

without sleep:
```
root@obstools-legacy:/code/obstools# ./devbox-helper-gcpscraper.sh
{"level":"debug","msg":"sinkWriter.Write(): wrote to outc chan","time":"2018-08-06T20:41:20Z"}
```

with sleep:
```
root@obstools-legacy:/code/obstools# ./devbox-helper-gcpscraper.sh
{"level":"debug","msg":"sinkWriter.Write(): wrote to outc chan","time":"2018-08-06T20:41:52Z"}
{"level":"debug","msg":"tcpStatsdSink.run(): received from outc chan","time":"2018-08-06T20:41:52Z"}
{"level":"debug","msg":"tcpStatsdSink.run(): wrote to conn","time":"2018-08-06T20:41:52Z"}
{"level":"debug","msg":"tcpStatsdSink.run(): updated lastFlushTime","time":"2018-08-06T20:41:52Z"}
{"level":"debug","msg":"tcpStatsdSink.run(): flushCond broadcasted","time":"2018-08-06T20:41:52Z"}
```

**Comment by mjchoi** [comment]
I've set up a branch on my application that you can use to reproduce:
* lyft/obstools#278

**Comment by jsedgwick** [comment]
PTAL @mjchoi @btc 

@mjchoi I tested with your repro, just inverted the error check...

**Comment by jsedgwick** [comment]
with the very latest commit Flush() could block forever but you probably have wayyyyyy bigger problems if you hit that (writing so many stats that that buffer channel never empties).

could work around it but IMO not worth code complexity

**Pull request merged**